pkg/networking/session_test.go

```diff
@@ -0,0 +1,483 @@
+	assert.NoError(t, err)
+
+	err = serverSession.Close()
+	assert.NoError(t, err)
+	err = serverSession.Close()
+	assert.NoError(t, err)
+}
+
+func TestOnClosedByOtherSide(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	mockProtocol := netmocks.NewMockProtocol(t)
+	mockProtocol.On("EmptyHandshake").Return(&textHandshake{}, nil)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("EmptyHeader").Return(&textHeader{}, nil)
+
+	clientHandler := netmocks.NewMockHandler(t)
+	serverHandler := netmocks.NewMockHandler(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clientConn, serverConn := testConnPipe()
+	net := networking.NewNetwork()
+
+	clientSession, err := net.NewSession(ctx, clientConn, testConfig(t, mockProtocol, clientHandler, "client"))
+	require.NoError(t, err)
+	serverSession, err := net.NewSession(ctx, serverConn, testConfig(t, mockProtocol, serverHandler, "server"))
+	require.NoError(t, err)
+
+	var closeWG sync.WaitGroup
+	closeWG.Add(1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	serverHandler.On("OnClose", serverSession).Return()
+	sc1 := serverHandler.On("OnHandshake", serverSession, &textHandshake{v: "hello"}).Once().Return()
+	sc1.Run(func(_ mock.Arguments) {
+		n, wErr := serverSession.Write([]byte("hello"))
+		assert.NoError(t, wErr)
+		assert.Equal(t, 5, n)
+		go func() {
+			// Close server after client received the handshake from server.
+			closeWG.Wait() // Wait for client to receive server handshake.
+			clErr := serverSession.Close()
+			assert.NoError(t, clErr)
+			wg.Done()
+		}()
+	})
+
+	clientHandler.On("OnClose", clientSession).Return()
+
+	// Send handshake to server.
+	n, err := clientSession.Write([]byte("hello"))
+	require.NoError(t, err)
+	assert.Equal(t, 5, n)
+
+	cs1 := clientHandler.On("OnHandshake", clientSession, &textHandshake{v: "hello"}).Once().Return()
+	cs1.Run(func(_ mock.Arguments) {
+		// On receiving handshake from server, signal to close the server.
+		closeWG.Done()
+		// Try to send message to server, but it will fail because server is already closed.
+		time.Sleep(10 * time.Millisecond) // Wait for server to close.
+		_, msgErr := clientSession.Write(encodeMessage("Hello session"))
+		require.ErrorIs(t, msgErr, io.ErrClosedPipe)
+		wg.Done()
+	})
+
+	wg.Wait() // Wait for client to finish.
+	err = clientSession.Close()
+	assert.ErrorIs(t, err, io.ErrClosedPipe) // Close reports the same error, because it was registered in the send loop.
+}
+
+func TestCloseParentContext(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	mockProtocol := netmocks.NewMockProtocol(t)
+	mockProtocol.On("EmptyHandshake").Return(&textHandshake{}, nil)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
+	mockProtocol.On("IsAcceptableHandshake", &textHandshake{v: "hello"}).Once().Return(true)
```

Comment:
Same lines

Reply:
Yep, I see that is intended.

Reply:
Yep, it's intended because `IsAcceptableHandshake` call happens two times (for client and for server).